content/sensu-core/1.0/guides/pre-compile-plugins.md

```diff
@@ -0,0 +1,146 @@
+---
+title: "Pre-Compiling Plugins"
+description: "Pre-Compiling Sensu Plugins for Distribution"
+product: "Sensu Core"
+version: "1.0"
+weight: 2
+next: ../intro-to-checks
+previous: ../overview
+menu:
+  sensu-core-1.0:
+    parent: guides
+---
+
+# Overview
+
+This guide will walk you through an example on how to pre-compile Sensu Plugins for distribution among clients.
+The goal of this guide is to showcase how you can quickly install plugins on ephemeral instances where time to compile can affect time to production  .
+We'll be using `sensu-plugins-aws` as it requires `nokogiri`,  usually requiring a longer build time.
+
+# Prerequist
+
+Sources:
+* [GitHub - luislavena/gem-compiler: A RubyGems plugin that generates binary gems](https://github.com/luislavena/gem-compiler)
+This guide assumes the following:
+* Centos 7 platform
+* You have some way to store these artifacts and make them available for systems in your environment
+_NOTE: Compiled gems are tied to the Ruby version they are compiled for. If Sensu upgrade Ruby 2.4 -> Ruby 2.5, packages will need to be rebuilt._
+_NOTE: Compiled gems are tied to CPU architecture for which they are compiled on._
+
+## Installing The Build Toolchain
+
+We'll first add the Sensu yum repository definition and install Sensu
+
+```bash
+printf '[sensu]
+name=sensu
+baseurl=https://repositories.sensuapp.org/yum/$releasever/$basearch/
+gpgcheck=0
+enabled=1' | sudo tee /etc/yum.repos.d/sensu.repo && yum install -y sensu
+```
+
+And then install EPEL and Dev tools
+
+```bash
+sudo yum install -y epel-release
+sudo yum groupinstall -y "Development Tools"
+```
+
+Finally we'll add Sensu's embedded bin to our path
+
+```bash
+export PATH=$PATH:/opt/sensu/embedded/bin
+```
+
+## Installing a Gem to a tmp directory
+
+Install the gem-compiler package
+
+```bash
+gem install gem-compiler
+```
+
+We'll now install the source gems into a cache directory
+
+```bash
+mkdir /tmp/gems
+gem install --no-ri --no-rdoc --install-dir /tmp/gems sensu-plugins-aws
+```
+
+Observing the `gem install` output, we can see that nokogiri and unf_ext build native extensions.
+These are the gems we will compile so we can install them from our own repository.
+
+### Compile the Gems for packaging
+
+We'll use `gem-compiler` to rebuild the gem packages with the included native extensions:
+
+```bash
+cd /tmp/gems/cache
+gem compile unf_ext-0.0.7.4.gem
+gem compile nokogiri-1.8.1.gem --prune
+```
+
+_NOTE: `nokogiri` requires --prune flag per gem-compiler readme notes _
+
+The resulting compiled gem files have a x86_64-linux in their file name denoting their system architech:
+
+```bash
+$ ls  -1 *-x86_64-linux.gem
+nokogiri-1.8.1-x86_64-linux.gem
```

Comment:
This also fails due to the version of nokogiri being `nokogiri-1.8.4.gem`. Let's not go after a specific version here.

Reply:
Add to this that Ruby devel is required. I'm getting failures when trying to compile noki:
```
# gem compile nokogiri-1.8.1.gem --prune
Unpacking gem: 'nokogiri-1.8.1' in temporary directory...
ERROR:  While executing gem ... (Gem::InstallError)
    invalid gem: No such file or directory @ rb_sysopen - nokogiri-1.8.1.gem
[root@sensu-enterprise-test cache]# gem compile nokogiri-1.8.4.gem --prune
Unpacking gem: 'nokogiri-1.8.4' in temporary directory...
Building native extensions.  This could take a while...
ERROR:  While executing gem ... (Gem::Ext::BuildError)
    ERROR: Failed to build gem native extension.

    current directory: /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/ext/nokogiri
/opt/sensu/embedded/bin/ruby -r ./siteconf20180911-22233-111auh8.rb extconf.rb
checking if the C compiler accepts ... yes
Building nokogiri using packaged libraries.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/opt/sensu/embedded/bin/$(RUBY_BASE_NAME)
        --help
        --clean
        --use-system-libraries
/opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:308:in `to_specs': Could not find 'mini_portile2' (~> 2.3.0) among 260 total gem(s) (Gem::MissingSpecError)
Checked in 'GEM_PATH=/root/.gem/ruby/2.4.0:/opt/sensu/embedded/lib/ruby/gems/2.4.0', execute `gem env` for more information
        from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:320:in `to_spec'
        from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_gem.rb:65:in `gem'
        from extconf.rb:463:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/lib/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4 for inspection.
Results logged to /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/lib/gem_make.out
```
This persists after installing ruby-devel via yum.

Reply:
added ruby devel and --use-system-libraries command, still shows a couple of warnings but builds successfully in my testing